pkg/container/versioned/value.go

```
@@ -0,0 +1,337 @@
+			if v.cleaner != nil {
+				v.cleanedVersion = keepVersion
+				v.cleaner(keepVersion)
+			} else {
+				log.WithFields(logrus.Fields{
+					"version": handle.getVersion(),
+					"handle":  name,
+				}).Error("closer: cleaner function not set, stale data will accumulate.")
+			}
+		}
+
+		v.mutex.Unlock()
+		return true
+	}
+}
+
+// versionRange is a range from the first to last version.
+// 'past' can be atomically modified to a smaller value.
+type versionRange struct {
+	first Version       // first version this value is valid for
+	past  atomicVersion // first version this value is invalid for
+}
+
+func (r *versionRange) Contains(version Version) bool {
+	return r.first <= version && version < r.past.Load()
+}
+
+type valueNode[T any] struct {
+	versions versionRange
+	next     atomic.Pointer[valueNode[T]]
+	value    T
+}
+
+// Value is a container for versioned value, implemented as a lock-free linked list
+type Value[T any] struct {
+	// versionedNodes are non-overlapping and in sorted order
+	head atomic.Pointer[valueNode[T]]
+}
+
+// SetValueAtVersion adds the value with validity starting from 'version'.
+// All values are added with "infinite" validity, which is then truncated
+// when an overlapping entry is added, or a value is removed.
+//
+// 'version' must be later than any the current version visible to the readers.
+// Must be called from a single writer!
+func SetValueAtVersion[T any](v *Value[T], value T, version Version) {
+	newNode := &valueNode[T]{
+		versions: versionRange{
+			first: version,
+		},
+		value: value,
+	}
+	newNode.versions.past.Store(invalidVersion)
+
+	// find if there is a current value that is valid for this new version
+	node := v.head.Load()
+	prev := &v.head
+	for node != nil {
+		if version < node.versions.first {
+			stacktrace := hclog.Stacktrace()
+			log.Errorf("AddValueAtVersion may not add values at versions lower than those already added! (%d < %d). Stacktrace: %s", version, node.versions.first, stacktrace)
```

I don't think "just" logging is a resilient way of dealing with this, why not return an error?

---

Ok, will add error return and check on the call sites :-)